File: bin/src/models/search_result.dart
import 'package:json_annotation/json_annotation.dart';

part 'search_result.g.dart';

@JsonSerializable()
class SearchResult {
  const SearchResult({
    required this.objectID,
    required this.categories,
    required this.content,
    required this.id,
    required this.permalink,
    required this.title,
    required this.version,
  });

  final String objectID;
  final List<String> categories;
  final String content;
  final String id;
  final String permalink;
  final String title;
  final num version;

  String get prettyTitle {
    if (id.startsWith('django.')) return id;
    if (id.startsWith('setting-')) return '$title [setting]';
    if (id.startsWith('templatefilter-')) return '$title [template filter]';
    if (id.startsWith('fieldlookup-')) return '$title [field lookup]';
    if (id.startsWith('templatetag-')) return '$title [template tag]';
    if (id.startsWith('cmdoption-')) return '$title [cmd option]';
    if (id.startsWith('envvar-')) return '$title [env var]';

    return title;
  }

  static const List<String> attributesToRetrieve = [
    'categories',
    'content',
    'id',
    'permalink',
    'title',
    'version',
  ];

  factory SearchResult.fromJson(Map<String, dynamic> json) =>
      _$SearchResultFromJson(json);

  Map<String, dynamic> toJson() => _$SearchResultToJson(this);
}

File: src/models/search_result.rs
use serde::Deserialize;
use serde_json::Number;

/// An Algolia result returned by the Django documentation index.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
pub struct SearchResult {
    /// Algolia object identifier used as Alfred's stable UID.
    #[serde(rename = "objectID")]
    pub object_id: String,
    /// Search index categories retained for compatibility with the provider payload.
    pub categories: Vec<String>,
    /// Searchable documentation content.
    pub content: String,
    /// Django documentation anchor or record identifier.
    pub id: String,
    /// Documentation URL opened by Alfred.
    pub permalink: String,
    /// Provider title, retained exactly as returned.
    pub title: String,
    /// Numeric provider version retained for complete historical deserialization.
    pub version: Number,
}

impl SearchResult {
    /// Applies Django's historical ID-prefix labels without changing the title text.
    pub fn pretty_title(&self) -> String {
        let suffix = if self.id.starts_with("django.") {
            return self.id.clone();
        } else if self.id.starts_with("setting-") {
            "setting"
        } else if self.id.starts_with("templatefilter-") {
            "template filter"
        } else if self.id.starts_with("fieldlookup-") {
            "field lookup"
        } else if self.id.starts_with("templatetag-") {
            "template tag"
        } else if self.id.starts_with("cmdoption-") {
            "cmd option"
        } else if self.id.starts_with("envvar-") {
            "env var"
        } else {
            return self.title.clone();
        };

        format!("{} [{suffix}]", self.title)
    }
}

/// Minimal subset of an Algolia single-index search response.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
pub struct SearchResponse {
    /// Search results in provider-defined ranking order.
    pub hits: Vec<SearchResult>,
}
